File: 4. MergeSort.cpp
/*4. Merge Sort

Implement Merge Sort using recursion.

Print the left and right subarrays during each merge.

Display the sorted subarray after each merge step.*/

#include <iostream>
#include <vector>
using namespace std;

// Function to print an array
void printArray(const vector<int> &arr, int start, int end) {
    cout << "[ ";
    for (int i = start; i <= end; i++)
        cout << arr[i] << " ";
    cout << "]";
}

// Merge two halves
void merge(vector<int> &arr, int left, int mid, int right) {
    cout << "\nMerging...\n";
    cout << "Left: ";
    printArray(arr, left, mid);
    cout << " | Right: ";
    printArray(arr, mid + 1, right);
    cout << endl;

    int n1 = mid - left + 1;
    int n2 = right - mid;

    vector<int> L(n1), R(n2);

    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0, j = 0, k = left;

    // Merge process
    while (i < n1 && j < n2) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    // Copy remaining elements
    while (i < n1)
        arr[k++] = L[i++];
    while (j < n2)
        arr[k++] = R[j++];

    cout << "Merged (Sorted Subarray): ";
    printArray(arr, left, right);
    cout << endl;
}

// Recursive Merge Sort
void mergeSort(vector<int> &arr, int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(arr, left, mid);      // Sort left half
        mergeSort(arr, mid + 1, right); // Sort right half
        merge(arr, left, mid, right);   // Merge halves
    }
}

int main() {
    vector<int> arr = {38, 27, 43, 3, 9, 82, 10};
    cout << "Original Array: ";
    for (int num : arr) cout << num << " ";
    cout << endl;
    mergeSort(arr, 0, arr.size() - 1);
    cout << "\nFinal Sorted Array: ";
    for (int num : arr) cout << num << " ";
    cout << endl;
    return 0;
}

/*

Time Complexity:

Best Case: O(n log n) → Array already sorted or any arrangement
Average Case: O(n log n) → Random order array
Worst Case: O(n log n) → Reverse sorted array

Space Complexity:
Auxiliary Space: O(n) → Needs extra array for merging

*/

File: 4. MergeSort.py
"""
Merge Sort Implementation in Python

This script demonstrates the recursive Merge Sort algorithm. It's a classic
divide-and-conquer algorithm known for its stable performance.
"""

def print_subarray(arr, start, end):
    """Helper function to print a slice of a list."""
    # Slicing creates the subarray, map converts numbers to strings, join combines them.
    sub = ' '.join(map(str, arr[start:end + 1]))
    print(f"[ {sub} ]", end="")

def merge(arr, left, mid, right):
    """Merges two sorted subarrays into a single sorted subarray."""
    print("\nMerging...")
    print("Left: ", end="")
    print_subarray(arr, left, mid)
    print(" | Right: ", end="")
    print_subarray(arr, mid + 1, right)
    print() # Newline

    # Create temporary lists for the left and right halves using slicing
    left_half = arr[left : mid + 1]
    right_half = arr[mid + 1 : right + 1]

    i = 0  # Initial index for the left half
    j = 0  # Initial index for the right half
    k = left # Initial index for the merged subarray

    # --- The Merge Process ---
    # Compare elements from both halves and copy the smaller one
    while i < len(left_half) and j < len(right_half):
        if left_half[i] <= right_half[j]:
            arr[k] = left_half[i]
            i += 1
        else:
            arr[k] = right_half[j]
            j += 1
        k += 1

    # --- Copy any remaining elements ---
    # If the left half has elements left
    while i < len(left_half):
        arr[k] = left_half[i]
        i += 1
        k += 1

    # If the right half has elements left
    while j < len(right_half):
        arr[k] = right_half[j]
        j += 1
        k += 1
        
    print("Merged (Sorted Subarray): ", end="")
    print_subarray(arr, left, right)
    print("\n" + "-"*30) # Separator for clarity

def merge_sort(arr, left, right):
    """
    The main recursive function that sorts arr[left..right] using Merge Sort.
    """
    if left < right:
        # Find the middle point to divide the array into two halves
        mid = (left + right) // 2

        # Recursively sort the first and second halves
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)

        # Merge the two sorted halves
        merge(arr, left, mid, right)

def main():
    """Main function to test the merge_sort."""
    arr = [38, 27, 43, 3, 9, 82, 10]
    
    print(f"Original Array: {' '.join(map(str, arr))}")
    print("-" * 30)
    
    merge_sort(arr, 0, len(arr) - 1)
    
    print("\nFinal Sorted Array: ", ' '.join(map(str, arr)))

if __name__ == "__main__":
    main()
    """
    Time Complexity:
    - Best Case: O(n log n)
    - Average Case: O(n log n)
    - Worst Case: O(n log n)

    Space Complexity:
    - Auxiliary Space: O(n) -> Requires extra space for temporary arrays during merging.
    
    """